= GTF
:gff-format: https://genome.ucsc.edu/FAQ/FAQformat.html#format3

GTF (Gene Transfer Format) is a refinement to {gff-format}[GFF^] that tightens the specification. The first eight GTF fields are the same as GFF. The group field has been expanded into a list of attributes. Each attribute consists of a `type`/`value` pair. Attributes must end in a semi-colon, and be separated from any following attribute by exactly one space.

The attribute list must begin with the two mandatory attributes:

[cols="10l,90"]
|===
| gene_id       | A globally unique identifier for the genomic source of the sequence.
| transcript_id | A globally unique identifier for the predicted transcript.
|===

.Ninth field in a GTF data line
[source,attributes]
----
gene_id "Em:U62317.C22.6.mRNA"; transcript_id "Em:U62317.C22.6.mRNA"; exon_number 1
----

.A whole GTF data line
[source,gtf]
----
chr12 <1>
ENSEMBL <2>
exon <3>
60059370 <4>
60061073 <5>
. <6>
- <7>
. <8>
gene_id "ENSMUSG00000020986"; transcript_id "ENSMUST00000021375"; exon_number "20"; gene_name "Sec23a"; gene_type "protein_coding"; transcript_name "Sec23a-001"; transcript_type "protein_coding"; <9>
----
<1> `seqname` - The name of the sequence. Must be a chromosome or scaffold.
<2> `source`  - The program that generated this feature.
<3> `feature` - The name of this type of feature. Some examples of standard feature types are `CDS`, `start_codon`, `stop_codon`, and `exon`.
<4> `start`   - The starting position of the feature in the sequence (`1-based`).
<5> `end`     - The ending position of the feature (inclusive).
<6> `score`   - A score between 0 and 1000.
<7> `strand`  - Valid entries include `+`, `-`, or `.` (for don't know/don't care).
<8> `frame`   - If the feature is a coding exon, frame should be a number between 0-2 that represents the reading frame of the first base. If the feature is not a coding exon, the value should be `.`.
<9> `group`   - All lines with the same group are linked together into a single item. Note the `gene_id` and `transcript_id` mandatory attributes.


== Feature types

The third field in the GFF specification represents feature type for the line. In addition to the standard features mentioned above there can be `intron`, `gene` and `transcript`.

.Other types of feature
[source,intron]
----
chr12   ENSEMBL intron  60079551        60083554        .       -       .       gene_id "ENSMUSG00000020986"; transcript_id "ENSMUST00000169976"; exon_number "1"; gene_name "Sec23a"; gene_type "protein_coding"; transcript_name "Sec23a-005"; transcript_type "retained_intron";
----
[source,gene]
----
chr12   ENSEMBL gene    60059370        60113004        .       -       .       gene_id "ENSMUSG00000020986"; transcript_id "ENSMUSG00000020986"; gene_type "protein_coding"; gene_status "NULL"; gene_name "Sec23a"; transcript_type "protein_coding"; transcript_status "NULL"; transcript_name "Sec23a";
----
[source,transcript]
----
chr12   ENSEMBL transcript      60073977        60083691        .       -       .       gene_id "ENSMUSG00000020986"; transcript_id "ENSMUST00000169976"; exon_number "4"; gene_name "Sec23a"; gene_type "protein_coding"; transcript_name "Sec23a-005"; transcript_type "retained_intron";
----
